app.py

```diff
@@ -36,52 +48,142 @@
+        return session
+    if quetz_token := channel_config.auth_quetz_token:
+        # use Quetz token auth
+        session.headers["X-API-Key"] = _unwrap_secret(quetz_token)
+        return session
+    if bearer_token := channel_config.auth_bearer_token:
+        # use bearer token auth
+        session.headers["Authorization"] = f"Bearer {_unwrap_secret(bearer_token)}"
+        return session
+    # no auth
+    return session
+
+
+
+@st.cache_resource(ttl="15m", max_entries=5)
+def rss_data(channel_name: str) -> ET.ElementTree | None:
+    """
+    :returns None if the channel does not have an RSS feed
+    """
+    channel_config = get_channel_config(channel_name)
+    if not channel_config.rss_enabled:
+        return None
+    rss_url = channel_config.rss_url
+    r = _make_http_session(channel_name).get(rss_url)
     r.raise_for_status()
     return ET.ElementTree(ET.fromstring(r.text))
 
 
-@st.cache_resource(ttl=FIFTEEN_MINS, max_entries=10)
-def channeldata(channel="conda-forge"):
-    if channel.startswith("pkgs/"):
-        r = requests.get(f"https://repo.anaconda.com/{channel}/channeldata.json")
-    else:
-        r = requests.get(f"https://conda.anaconda.org/{channel}/channeldata.json")
+@st.cache_resource(ttl="15m", max_entries=10)
+def get_channeldata(channel_name: str) -> dict:
+    r = _make_http_session(channel_name).get(get_channel_config(channel_name).channeldata_url)
     r.raise_for_status()
     return r.json()
 
 
-@st.cache_resource(ttl=FIFTEEN_MINS, max_entries=1000)
-def api_data(package_name, channel="conda-forge"):
-    if channel.startswith("pkgs/"):
-        channel = channel.split("/", 1)[1]
-    r = requests.get(f"https://api.anaconda.org/package/{channel}/{package_name}/files")
+def _download_compressed_repodata(channel_name: str, arch_subdir: str) -> dict | None:
+    """
+    Try to download the compressed repodata.json file.
+    If the file does not exist, return None.
+    Other HTTP errors are raised.
+    :returns the decompressed repodata.json, or None if the compressed file does not exist
+    """
+    channel_config = get_channel_config(channel_name)
+    zstd_url = channel_config.get_zstd_repodata_url(arch_subdir)
+
+    with _make_http_session(channel_name).get(zstd_url, stream=True) as r:
+        if r.status_code == 404:
+            return None
+        r.raise_for_status()
+
+        dctx = zstd.ZstdDecompressor()
+        with dctx.stream_reader(r.raw) as reader:
+            return json.load(reader)
+
+@st.cache_resource(ttl="15m", max_entries=50)
+def get_repodata(channel_name: str, arch_subdir: str) -> dict:
+    """
+    Fetch the repodata.json for a channel and subdir.
+    This function first tries to download the .zst-compressed version of the repodata, and if that does not exist,
+    it falls back to the uncompressed.
+    """
+    decompressed_repodata = _download_compressed_repodata(channel_name, arch_subdir)
+    channel_config = get_channel_config(channel_name)
```

Comment:
This line can be moved after the early return below, I think.

Reply:
e3e89e06f10d2c0c1d7cbe8bdcd4df8505a066ba